src/blocks.jl

```diff
@@ -0,0 +1,95 @@
+"""
+    blocks(a::AbstractArray{T,N}) :: AbstractArray{<:AbstractArray{T,N},N}
+
+Return the array-of-arrays view to `a` such that
+
+```
+blocks(a)[i₁, i₂, ..., iₙ] == a[Block(i₁), Block(i₂), ..., Block(iₙ)]
+```
+
+This function does not copy the blocks and give a mutable viwe to the original
+array.  This is an "inverse" of [`mortar`](@ref).
+
+# Examples
+```jldoctest; setup = quote using BlockArrays end
+julia> bs1 = permutedims(reshape([
+               1ones(1, 3), 2ones(1, 2),
+               3ones(2, 3), 4ones(2, 2),
+           ], (2, 2)))
+2×2 Array{Array{Float64,2},2}:
+ [1.0 1.0 1.0]               [2.0 2.0]
+ [3.0 3.0 3.0; 3.0 3.0 3.0]  [4.0 4.0; 4.0 4.0]
+
+julia> a = mortar(bs1)
+2×2-blocked 3×5 BlockArray{Float64,2}:
+ 1.0  1.0  1.0  │  2.0  2.0
+ ───────────────┼──────────
+ 3.0  3.0  3.0  │  4.0  4.0
+ 3.0  3.0  3.0  │  4.0  4.0
+
+julia> bs2 = blocks(a)
+2×2 Array{Array{Float64,2},2}:
+ [1.0 1.0 1.0]               [2.0 2.0]
+ [3.0 3.0 3.0; 3.0 3.0 3.0]  [4.0 4.0; 4.0 4.0]
+
+julia> bs1 == bs2
+true
+
+julia> bs2[1, 1] .*= 100;
+
+julia> a  # in-place mutation is reflected to the block array
+2×2-blocked 3×5 BlockArray{Float64,2}:
+ 100.0  100.0  100.0  │  2.0  2.0
+ ─────────────────────┼──────────
+   3.0    3.0    3.0  │  4.0  4.0
+   3.0    3.0    3.0  │  4.0  4.0
+```
+"""
+blocks(a::AbstractArray) = blocks(PseudoBlockArray(a, axes(a)))
+blocks(a::AbstractBlockArray) = BlocksView(a)
+blocks(a::BlockArray) = a.blocks
+blocks(A::Adjoint) = adjoint(blocks(parent(A)))
+blocks(A::Transpose) = transpose(blocks(parent(A)))
```

Comment:
I'm a bit puzzled why you needed `transpose`/`adjoint` twice in https://github.com/JuliaArrays/BlockArrays.jl/pull/119/commits/7ff342154a9cab1e91845bc0c6c398a4023313e1

Isn't doing this in the outer-most matrix enough, since `adjoint` and `transpose` are recursive?  I think this is the way to do this if we want the property `blocks(a)[i₁, i₂, ..., iₙ] == a[Block(i₁), Block(i₂), ..., Block(iₙ)]`.  At least it looks like this test is working:

https://github.com/JuliaArrays/BlockArrays.jl/blob/882ba4aff84daa5e65af035db2b8f28d7988292a/test/test_blocks.jl#L71-L75

Reply:
Good poibt